XamarinCommunityToolkit/Effects/EntryFormattedTextEffect.android.cs

```diff
@@ -0,0 +1,258 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using Android.Text;
+using Android.Text.Style;
+using Android.Widget;
+using Xamarin.CommunityToolkit.Android.Effects;
+using Xamarin.CommunityToolkit.Effects;
+using Xamarin.Forms;
+using Xamarin.Forms.Platform.Android;
+
+[assembly: ResolutionGroupName(nameof(Xamarin.CommunityToolkit))]
+[assembly: ExportEffect(typeof(EntryFormattedTextEffect), nameof(EntryFormattedTextEffect))]
+
+namespace Xamarin.CommunityToolkit.Android.Effects
+{
+	/// <summary>
+	/// The FormattedText <see cref="PlatformEffect"/> specific
+	/// to <see cref="Entry"/> controls.
+	/// </summary>
+	public class EntryFormattedTextEffect : PlatformEffect, IDisposable
+	{
+		bool isDisposed;
+
+		protected IList<FormattedTextSegment<Java.Lang.Object>> Segments { get; set; }
+
+		/// <summary>
+		/// Gets the native <see cref="FormsEditText"/> control.
+		/// </summary>
+		public new FormsEditText Control => (FormsEditText)base.Control;
+
+		/// <summary>
+		/// Gets the Xamarin.Forms <see cref="Entry"/> element.
+		/// </summary>
+		public new Entry Element => (Entry)base.Element;
+
+		/// <inheritdoc />
+		protected override void OnAttached()
+		{
+			UpdateFormattedText();
+
+			Control.TextChanged += OnTextChanged;
+		}
+
+		/// <inheritdoc />
+		protected override void OnDetached()
+		{
+		}
+
+		/// <inheritdoc />
+		protected override void OnElementPropertyChanged(PropertyChangedEventArgs args)
+		{
+			base.OnElementPropertyChanged(args);
+
+			if (args.PropertyName == FormattedTextEffect.FormattedTextProperty.PropertyName)
+				UpdateFormattedText();
+		}
+
+		void OnTextChanged(object sender, global::Android.Text.TextChangedEventArgs args)
+		{
+			var spannable = args.Text as SpannableStringBuilder;
+			if (spannable == null || Segments.Count == 0)
+			{
+				return;
+			}
+
+			var segment = Segments[args.Start];
+			var length = args.AfterCount - args.BeforeCount;
+			var startIndex = args.Start - segment.StartIndex;
+
+			if (length > 0)
+			{
+				segment.Span.Text = segment.Span.Text.Insert(startIndex, Control.Text.Substring(args.Start, length));
+
+				for (var index = 0; index < length; index++)
+				{
+					Segments.Insert(args.Start + index, segment);
+				}
+			}
+			else if (length < 0)
+			{
+				var absLength = Math.Abs(length);
+				segment.Span.Text = segment.Span.Text.Remove(startIndex, Math.Abs(length));
+
+				for (var index = 0; index < absLength; index++)
+				{
+					Segments.RemoveAt(args.Start + index);
+				}
+			}
+
+			var position = 0;
+			foreach (var item in Segments.Distinct())
+			{
+				item.StartIndex = position;
+				position += item.Span.Text.Length;
+			}
+
+			foreach (var nativeSpan in segment.NativeSpans)
+			{
+				spannable.RemoveSpan(nativeSpan);
+				spannable.SetSpan(nativeSpan, segment.StartIndex, segment.StartIndex + segment.Span.Text.Length, SpanTypes.ExclusiveExclusive);
+			}
+		}
```

Comment:
This algorithm has been re-written from the original PR with a focus on managing the state via shared code and only applying the spans to native objects when needed. I am hoping we can extract the rules in this algorithm to use a shared object across the different platform implementations. If we can do that it means the business rules will be consistent and much easier to maintain then N-number of implementations.

The rules are documented in the linked issue